docs/proposals/decouple-application-sync-user-using-impersonation

```diff
@@ -0,0 +1,481 @@
+- If impersonation feature is enabled, and no service account name is provided in the Application CR, then it should use a default service account of the `Application` namespace.
+- Access restrictions implemented through properties in AppProject (if done) must have the existing behavior. From a security standpoint, any restrictions that were available before switching to a service account based approach should continue to exist even when the impersonation feature is enabled.
+
+### Non-Goals
+
+What is out of scope for this proposal? Listing non-goals helps to focus discussion and make progress.
+None
+
+## Proposal
+
+Enable ArgoCD Admin to specify a `serviceAccountName` in `AppProjects` CR on a per `destination` basis.
+When applications gets synced, based on its destination, the `serviceAccountName` configured in the `AppProject` will be chosen and used for impersonation when executing the kubectl apply/create commands. There could either be a regex based match of the destination or an exact match of the destination. If there are multiple possible options, then an error would be thrown during the sync operation.
+
+```
+apiVersion: argoproj.io/v1alpha1
+kind: AppProject
+metadata:
+  name: my-project
+  namespace: argocd
+  finalizers:
+    - resources-finalizer.argocd.argoproj.io
+spec:
+  description: Example Project
+  # Allow manifests to deploy from any Git repos
+  sourceRepos:
+  - '*'
+  destinations:
+  - namespace: guestbook
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-deployer
+  - namespace: guestbook-dev
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-dev-deployer
+  - namespace: guestbook-stage
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-stage-deployer
+```
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+#### Use case 1:
+
+As a user, I would like to use kubernetes security constructs to restrict user access for application sync
+So that, I can provide granular permissions based on the principal of least privilege required for syncing an application.
+
+#### Use case 2:
+
+As a user, I would like to configure a common service account for all applications associated to an AppProject
+So that, I can use a generic convention of naming service accounts and avoid associating the service account per application.
+
+### Implementation Details/Notes/Constraints
+
+#### Component : GitOps Engine
+
+- Fix GitOps Engine code to honor Impersonate configuration set in the Application sync context for all kubectl commands that are being executed.
+
+#### Component: ArgoCD API
+
+- Support new property `serviceAccountName` in `ApplicationDestination` section of `AppProject`
```

Comment:
To leave it flexible, instead of `serviceAccountName` I'd just expose the `as` and `asGroup` attributes, to match kubectl flags and kubecontext configuration.